### PR TITLE
feat: Use criteria builder(RWA-1356)

### PR DESCRIPTION
**Pull request by dharmendrak**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-1356

### Change description ###

Use criteria builder api

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/TaskResourceQueryBuilder.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Order;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+
+public abstract class TaskResourceQueryBuilder<T> {
+    protected final EntityManager entityManager;
+    protected final CriteriaBuilder builder;
+    protected final CriteriaQuery<T> query;
+    protected final Root<TaskResource> root;
```
Comment (rasikasri) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/TaskResourceQueryBuilder.java:18]: This need to be generic type T, isn't it?. When you declare it as TaskResource, then TaskResourceQueryBuilder can only be used with TaskResource.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/entities/TaskResourceSummary.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package uk.gov.hmcts.reform.wataskmanagementapi.cft.entities;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+import java.time.OffsetDateTime;
+
+@Getter
+@AllArgsConstructor
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/entities/TaskResourceSummary.java:9]: We generally avoid lombok annotations that can be easily generated.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/entities/TaskResourceSummary.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+public class TaskResourceSummary {
+
+    private String taskId;
+    private OffsetDateTime dueDateTime;
+    private String caseId;
+    private String caseName;
+    private String caseCategory;
+    private String locationName;
+    private String title;
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/entities/TaskResourceSummary.java:18]: This is probably a future improvement if there's any performance gain but I will note it here.
For each search we will know the sorting parameters from the request.  What we allow is only 1 sorting parameter and then apply the "default" sorting as a secondary (which is currently by dueDate).

With that in mind we _could_ reduce the columns we retrieve to just taskId, thePrimarySortColumn, dueDate.

Not sure if that improves the performance much but every small gain may add up.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java [hunk 4]
```diff
@@ -65,39 +75,81 @@ public GetTasksResponse<Task> searchForTasks(
...
+    }
+
+    private List<TaskResourceSummary> getTaskResourceSummary(int firstResult,
+                                                             int maxResults,
+                                                             SearchTaskRequest searchTaskRequest,
+                                                             AccessControlResponse accessControlResponse,
+                                                             List<PermissionTypes> permissionsRequired) {
+        Sort sort = SortQuery.sortByFields(searchTaskRequest);
+        Pageable page = OffsetPageableRequest.of(firstResult, maxResults, sort);
+        TaskResourceSummaryQueryBuilder summaryQueryBuilder = new TaskResourceSummaryQueryBuilder(entityManager);
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:113]: If all we do is pass the entityManager elsewhere do we need it in here or is it better to have it injected into where we are passing it?

Comment (raghera) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:113]: I think it's ok for now as I think we most likly want to reorganise/refactor later anyway.

**Review by raghera — COMMENTED** [review]
Looks good, mainly minor comments.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java (hunk 4 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:113]: I have started refactoring in another PR

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/SortQuery.java [hunk 5]
```diff
@@ -43,6 +63,21 @@ private static List<Sort> generateSortList(List<SortingParameter> sortingParamet
         }).filter(Objects::nonNull).collect(Collectors.toList());
     }
 
+    private static List<Order> generateOrders(List<SortingParameter> sortingParameters,
+                                              CriteriaBuilder builder,
+                                              Root<TaskResource> root) {
+        return sortingParameters.stream().map((sortingParameter) -> {
+            switch (sortingParameter.getSortOrder()) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/SortQuery.java:70]: New line following re-engineering for performance. Can we add unit test to cover this line?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/TaskResourceSummaryQueryBuilder.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+                                                root.get("dueDateTime"),
+                                                root.get("caseId"),
+                                                root.get("caseName"),
+                                                root.get("caseCategory"),
+                                                root.get("locationName"),
+                                                root.get("title"));
+
+        CriteriaQuery<TaskResourceSummary> criteriaQuery =  query.multiselect(selections).distinct(true);
+
+        if (predicate != null) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/TaskResourceSummaryQueryBuilder.java:34]: New line following re-engineering for performance. Can we add unit test to cover this line?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/TaskResourceSummaryQueryBuilder.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            criteriaQuery.where(predicate);
+        }
+        if (orders != null && !orders.isEmpty()) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/TaskResourceSummaryQueryBuilder.java:37]: New line following re-engineering for performance. Can we add unit test to cover this line?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java [hunk 8]
```diff
@@ -309,18 +319,21 @@ private static Predicate searchByRegion(Root<TaskResource> root,
...
     private static Predicate searchByIncludingCaseId(Root<TaskResource> root,
                                                      CriteriaBuilder builder,
-                                                     RoleAssignment roleAssignment) {
-        String caseId = roleAssignment.getAttributes().get(CASE_ID.value());
-        if (StringUtils.hasText(caseId)) {
-            return builder.equal(root.get(CASE_ID_COLUMN), caseId);
+                                                     RoleAssignmentForSearch roleAssignment) {
+        Set<String> caseIds = roleAssignment.getCaseIds();
+        if (!CollectionUtils.isEmpty(caseIds)) {
+            if (caseIds.size() == ONE) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java:325]: New line following re-engineering for performance. Can we add unit test to cover this line?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java [hunk 9]
```diff
@@ -309,18 +319,21 @@ private static Predicate searchByRegion(Root<TaskResource> root,
...
+                return builder.equal(root.get(CASE_ID_COLUMN), caseIds.iterator().next());
+            }
+            return builder.in(root.get(CASE_ID_COLUMN)).value(caseIds);
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java:328]: New line following re-engineering for performance. Can we add unit test to cover this line?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java [hunk 10]
```diff
@@ -241,9 +251,9 @@ private static Predicate getEmptyOrNullAuthorizationsPredicate(
     private static Predicate mapSecurityClassification(
         Root<TaskResource> root,
         CriteriaBuilder builder,
-        RoleAssignment roleAssignment) {
-        final Classification classification = roleAssignment.getClassification();
-        if (classification != null) {
+        RoleAssignmentForSearch roleAssignment) {
+        if (roleAssignment.getClassification() != null) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentFilter.java:255]: New line following re-engineering for performance. Can we add unit test to cover this line?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java [hunk 11]
```diff
@@ -133,11 +211,50 @@ public Optional<TaskResource> getTask(String taskId,
...
+
+        selectQueryBuilder.where(selectPredicate).build().getResultList();
+
+        try {
+            return selectQueryBuilder
+                .where(selectPredicate)
+                .build()
+                .getResultList().stream()
+                .findFirst();
+        } catch (NoResultException ne) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/CftQueryService.java:233]: New line following re-engineering for performance. Can we add unit test to cover this line (edge case negative scenarios)?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentSearchData.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            return Collections.emptySet();
+        }
+
+        //Group the Case roles
+        Map<Integer, List<RoleAssignmentForSearch>> groupedCaseRoles = caseRoleAssignments
+            .stream()
+            .collect(Collectors.groupingBy(RoleAssignmentForSearch::hashCode));
+
+        return groupedCaseRoles.values().stream()
+            .filter(caseRoles -> !CollectionUtils.isEmpty(caseRoles))
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/RoleAssignmentSearchData.java:63]: New line following re-engineering for performance. Can we add unit test to cover this line?

**Review by BenNzewi — APPROVED** [review]

**Review by serdaroguzhanakin — APPROVED** [review]
